### PR TITLE
feat: configurable fee collector for DeductFeeDecorator

### PR DESCRIPTION
**Pull request by JimLarson**
## Description

Refs: Agoric/agoric-sdk#9036

Make the fee collector module name in `DeductFeeDecorator` configurable.

Pulls a customization from Agoric/agoric-sdk into agoric-labs/cosmos-sdk so it can evolve along with cosmos-sdk.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [x] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [x] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by JimLarson** [comment]
Lint failure expected.

**Comment by JimLarson** [comment]
Should be cherry-picked onto a "release" branch. Following past precedent, which it's unnecessary to improve upon, we'll branch `Agoric-v0.46.16.2.x` off of `v0.46.16-alpha.agoric.2` and lay down the tag `v0.46.16-alpha.agoric.2.1`.

File: x/auth/ante/fee.go [hunk 1]
```diff
@@ -17,22 +17,30 @@ type TxFeeChecker func(ctx sdk.Context, tx sdk.Tx) (sdk.Coins, int64, error)
...
+	feegrantKeeper   FeegrantKeeper
+	txFeeChecker     TxFeeChecker
+	feeCollectorName string
 }
 
 func NewDeductFeeDecorator(ak AccountKeeper, bk types.BankKeeper, fk FeegrantKeeper, tfc TxFeeChecker) DeductFeeDecorator {
+	return NewDeductFeeDecoratorWithName(ak, bk, fk, tfc, types.FeeCollectorName)
+}
+
+// NewDeductFeeDecorator returns a DeductFeeDecorator using a custom fee collector module account.
```
Comment (gibson042) [inline x/auth/ante/fee.go]:
```suggestion
// NewDeductFeeDecoratorWithName returns a DeductFeeDecorator using a custom fee collector module account.
```

**Review by gibson042 — APPROVED** [review]

File: x/auth/ante/fee_test.go [hunk 2]
```diff
@@ -146,3 +147,47 @@ func (s *AnteTestSuite) TestDeductFees() {
...
+	gasLimit := testdata.NewTestGasLimit()
+	s.Require().NoError(s.txBuilder.SetMsgs(msg))
+	s.txBuilder.SetFeeAmount(feeAmount)
+	s.txBuilder.SetGasLimit(gasLimit)
+
+	privs, accNums, accSeqs := []cryptotypes.PrivKey{priv1}, []uint64{0}, []uint64{0}
+	tx, err := s.CreateTestTx(privs, accNums, accSeqs, s.ctx.ChainID())
+	s.Require().NoError(err)
+
+	// Set account with sufficient funds
```
Comment (gibson042) [inline x/auth/ante/fee_test.go]:
```suggestion
	// Set transacting account with sufficient funds
```

File: x/auth/ante/fee_test.go [hunk 3]
```diff
@@ -146,3 +147,47 @@ func (s *AnteTestSuite) TestDeductFees() {
...
+	altCollectorName := "distribution"
+	altCollectorAcc := s.app.AccountKeeper.GetModuleAccount(s.ctx, altCollectorName)
+	s.Require().True(s.app.BankKeeper.GetAllBalances(s.ctx, feeCollectorAcc.GetAddress()).Empty())
+	altBalance := s.app.BankKeeper.GetAllBalances(s.ctx, altCollectorAcc.GetAddress())
+
+	dfd := ante.NewDeductFeeDecoratorWithName(s.app.AccountKeeper, s.app.BankKeeper, nil, nil, altCollectorName)
+	antehandler := sdk.ChainAnteDecorators(dfd)
+
+	_, err = antehandler(s.ctx, tx, false)
+	s.Require().NoError(err)
```
Comment (gibson042) [inline x/auth/ante/fee_test.go]:
```suggestion
	// Run the transaction through a handler chain that deducts fees into altCollectorAcc. 
	dfd := ante.NewDeductFeeDecoratorWithName(s.app.AccountKeeper, s.app.BankKeeper, nil, nil, altCollectorName)
	antehandler := sdk.ChainAnteDecorators(dfd)
	_, err = antehandler(s.ctx, tx, false)
	s.Require().NoError(err)
```

File: x/auth/ante/fee_test.go [hunk 4]
```diff
@@ -146,3 +147,47 @@ func (s *AnteTestSuite) TestDeductFees() {
...
+
+	s.Require().True(s.app.BankKeeper.GetAllBalances(s.ctx, feeCollectorAcc.GetAddress()).Empty())
+	newAltBalance := s.app.BankKeeper.GetAllBalances(s.ctx, altCollectorAcc.GetAddress())
+	s.Require().True(newAltBalance.IsAllGTE(altBalance))
```
Comment (gibson042) [inline x/auth/ante/fee_test.go:192]:
```suggestion
	// Assert that feeCollectorAcc is still empty but the altCollectorAcc balance has increased.
	newFeeCollectorBalance := s.app.BankKeeper.GetAllBalances(s.ctx, feeCollectorAcc.GetAddress())
	s.Require().True(newFeeCollectorBalance.Empty())
	newAltBalance := s.app.BankKeeper.GetAllBalances(s.ctx, altCollectorAcc.GetAddress())
	s.Require().True(newAltBalance.IsAllGTE(altBalance))
	s.Require().False(newAltBalance.IsEqual(altBalance))
```

File: x/auth/ante/fee.go (hunk 1 above)
Comment (JimLarson) [reply x/auth/ante/fee.go]: Oops, done.

File: x/auth/ante/fee_test.go (hunk 2 above)
Comment (JimLarson) [reply x/auth/ante/fee_test.go]: Done.

File: x/auth/ante/fee_test.go (hunk 3 above)
Comment (JimLarson) [reply x/auth/ante/fee_test.go]: Done.

File: x/auth/ante/fee_test.go (hunk 4 above)
Comment (JimLarson) [reply x/auth/ante/fee_test.go:192]: Done.

**Pull request merged**